crates/interpreter/src/instructions/opcode.rs

```diff
@@ -1,454 +1,314 @@
-use crate::gas;
-use crate::primitives::SpecId;
+//! EVM opcode definitions and utilities.
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+use super::*;
+use crate::{
+    gas,
+    primitives::{Spec, SpecId},
+    Host, Interpreter,
+};
+use core::fmt;
+
+macro_rules! opcodes {
+    ($($val:literal => $name:ident => $f:expr),* $(,)?) => {
+        // Constants for each opcode. This also takes care of duplicate names.
+        $(
+            #[doc = concat!("The `", stringify!($val), "` (\"", stringify!($name),"\") opcode.")]
+            pub const $name: u8 = $val;
+        )*
+
+        /// Maps each opcode to its name.
+        pub const OPCODE_JUMPMAP: [Option<&'static str>; 256] = {
+            let mut map = [None; 256];
+            let mut prev: u8 = 0;
+            $(
+                let val: u8 = $val;
+                assert!(val == 0 || val > prev, "opcodes must be sorted in ascending order");
+                prev = val;
+                map[$val] = Some(stringify!($name));
+            )*
+            let _ = prev;
+            map
+        };
+
+        type Instruction = fn(&mut Interpreter, &mut dyn Host);
+
+        // Requires `inline_const` and `const_mut_refs` unstable features,
+        // but provides ~+2% extra performance.
+        // See: https://github.com/bluealloy/revm/issues/310#issuecomment-1664381513
+        /*
+        type InstructionTable = [Instruction; 256];
+
+        const fn make_instruction_table<SPEC: Spec>() -> InstructionTable {
+            let mut table: InstructionTable = [control::not_found; 256];
+            let mut i = 0usize;
+            while i < 256 {
+                table[i] = match i as u8 {
+                    $($name => $f,)*
+                    _ => control::not_found,
+                };
+                i += 1;
+            }
+            table
+        }
+
+        // in `eval`:
+        (const { make_instruction_table::<SPEC>() })[opcode as usize](interpreter, host)
+        */
+
+        /// Evaluates the opcode in the given context.
+        #[inline(always)]
+        pub(crate) fn eval<SPEC: Spec>(opcode: u8, interpreter: &mut Interpreter, host: &mut dyn Host) {
+            // See https://github.com/bluealloy/revm/issues/310#issuecomment-1664381513
+            // for previous efforts on optimizing this function.
+            let f: Instruction = match opcode {
+                $($name => $f as Instruction,)*
+                _ => control::not_found as Instruction,
```

Comment:
I like this a lot! casting to `Instruction` allows us to be sure that every `fn` has the same signature. @DaniPopes do you think this is the reason for the performance boost?

Reply:
IIRC it's a bit more than half of the total perf improvement of this pr